Improved Settings

sarathms opened the pull request:
Fixes ooni/probe#1088
Fixes ooni/probe#1022
Fixes ooni/probe#1121

- [x] Refactor settings components as functional components
- [x] Update layout to new mockups in ooni/probe#1125
- [x] Add a LangaugeSelector MVP
- [x] Show supported languages in selector dropdown

hellais commented:
I see that when I change the language there are some strings that have still boilerplate in them, why is that?
![Screenshot 2020-09-01 at 10 17 32](https://user-images.githubusercontent.com/424620/91825471-a04fd200-ec3c-11ea-8dca-774f1aeb2bcc.png)

hellais commented:
The category code selection needs a bit more work:

![Screenshot 2020-09-01 at 10 18 33](https://user-images.githubusercontent.com/424620/91825640-d1c89d80-ec3c-11ea-9d2f-bf0ba28e3574.png)

- The top right corner close icon should receive the same treatment as the modal in the stop test UI
- The subtle fade effect doesn't work well and I think it's affecting the buttons as well, let's remove it and document it as future work

- At the end of the page there is unnecessary whitespace:
![Screenshot 2020-09-01 at 10 20 02](https://user-images.githubusercontent.com/424620/91825761-f9b80100-ec3c-11ea-9ff1-492a3540cd2a.png)

- When I close the modal after having made some changes, I would expect them to not be saved, while in reality they are saved

hellais reviewed (changes requested):
Needs a bit more polishing. Left some feedback in comments.

sarathms commented:
![image](https://user-images.githubusercontent.com/700829/98380977-480bc680-2017-11eb-93e8-774615f8f6b4.png)

sarathms commented:
@hellais I refactored the modal code to share the code between this screen and the 'stop-the-test' screen. Please test that screen also a few times along with this.

@agrabeli Can you review/edit [the copy](https://docs.google.com/spreadsheets/d/1GPkr_OyNhXJRTXnseyNbo2P4Du-i_W9kJdqjZG4HVkM/edit#gid=0) for this confirmation popup? Keys are like `Settings.Websites.Categories.*`. If/when you edit it, do notify me here or on slack, so that I can include those changes to this PR. Thanks!

sarathms commented:
@hellais Please note that I changed how we write the categories into the config file. Instead of an array, it is a dictionary/map. Reasons : *) will be easier for someone trying to edit the file manually *) made it easier to write some code
```
{
  "_version": 3,
  "_informed_consent": true,
  "sharing": {
    ...
  },
  "nettests": {
    "websites_url_limit": 10,
    "websites_enabled_category_codes": {
      "ALDR": true,
      "ANON": true,
      "COMM": false,
      "COMT": true,
      "CTRL": false,
```
On a related note, in this change we are bumping the config file version because we added the categories entry.

hellais commented on main/utils/config.js:
```diff
@@ -5,28 +5,62 @@ const log = require('electron-log')
 
 const OONI_CONFIG_PATH = path.join(getHomeDir(), 'config.json')
 
+const availableCategoriesList = {
```
This should be a list

hellais commented on main/utils/config.js:
```diff
@@ -5,28 +5,62 @@ const log = require('electron-log')
-      "websites_url_limit": 0
+    'nettests': {
+      'websites_url_limit': 0,
+      'websites_enabled_category_codes': availableCategoriesList
```
You can do something like: 
```
Object.keys(availableCategoriesList).reduce(k => availableCategoriesList[k] === true)
```

hellais commented:
This is starting to look good!

![Screenshot 2020-11-09 at 11 58 20](https://user-images.githubusercontent.com/424620/98533220-f9fafb00-2282-11eb-9f3f-73fa2f2ef6e5.png)

A few small things that I think should be addressed:
- The scroll bar is scrollable in the background of the modal (see above screenshot). It should instead not be scrollable when the modal is displayed.
- Unsaved changes are displayed even if no change was made
- The unsaved changes modal should replace the current modal and not be displayed as an overlay
- The fonts in the modal look a bit weird. Could it be that some CSS effect is applied to them?

hellais commented:
I did some light testing of it and functionally it's working as expected.

sarathms commented:
> The unsaved changes modal should replace the current modal and not be displayed as an overlay

@hellais Isn't it a common pattern to show this kind of save-or-discard confirmation prompt before taking down the UI where the changes were made? It seems counter-intuitive despite the fact it is a modal over a modal.

hellais commented:
> Isn't it a common pattern to show this kind of save-or-discard confirmation prompt before taking down the UI where the changes were made? 

It's generally considered an anti-pattern to show a modal within another modal, see:
* https://ux.stackexchange.com/questions/52042/is-it-acceptable-to-open-a-modal-popup-on-top-of-another-modal-popup
* https://ux.stackexchange.com/questions/120182/display-modal-over-another-modal-its-really-a-sin

If we really want to keep the modal we should either:
* Remove the one below and just show the confirm modal
* Keep both, but make the second focussed by fading out the one in the background by putting it in the same faded out background

I don’t think it’s really needed to have a confirm are you sure to cancel modal, since it’s not really such an important thing to save. Perhaps we can make it clearer that the done button needs to be clicked on in order to save the data, by renaming it from DONE to SAVE.

That way it’s clear that the X is to cancel, while the DONE button is needed in order to commit the state to the disk.

sarathms commented:
@hellais What exactly is the problem with fonts? I am yet to notice anything. I checked for CSS effects or overrides, nothing applies. All the text that I checked has the right fonts applied.

hellais commented:
It's hard to describe. They look a bit "fuzzy" like there is some odd aliasing effect going on, but maybe it's an optical illusion or a brainfart on my end.

sarathms commented:
Category Code Selections
![image](https://user-images.githubusercontent.com/700829/99098974-c502e700-25a7-11eb-895c-233b61dd2524.png)

Reordered sections
![image](https://user-images.githubusercontent.com/700829/99091555-0e4e3900-259e-11eb-8689-2b2a60abb393.png)

Localized names of languages
![image](https://user-images.githubusercontent.com/700829/99098887-a69ceb80-25a7-11eb-9a7b-bd756f4cb35b.png)

hellais reviewed (approved):
LGTM
